signal.c

```diff
@@ -400,9 +400,14 @@ evsig_handler(int sig)
 #ifdef _WIN32
 	send(evsig_base_fd, (char*)&msg, 1, 0);
 #else
-	{
-		int r = write(evsig_base_fd, (char*)&msg, 1);
-		(void)r; /* Suppress 'unused return value' and 'unused var' */
+	for (;;) {
+		errno = ENOSPC; /* for 0 */
```

Comment:
What for?

Reply:
For the case where write fails to write the byte and returns 0. This is allowed; write does not guarantee that the entire requested block is written. In this situation write probably should not be retried, since it may just continue returning 0 and create an infinite loop. Thus, set errno to describe what happened, log a warning, and leave. This will result in a missing signal, but seems to be the best solution.

Reply:
>For the case where write fails to write the byte and returns 0.

But in this case errno should be set to EAGAIN, am I missing something?

Reply:
EAGAIN is only returned for non-blocking fds. When write does a partial write it returns the number of bytes written and does not set errno.

Reply:
This is non-blocking fd

https://github.com/libevent/libevent/blob/8e03f4959736731975bf0eb0433e00486910ade6/signal.c#L183

https://github.com/libevent/libevent/blob/8e03f4959736731975bf0eb0433e00486910ade6/evutil.c#L2874-L2892

Reply:
Ah, I didn't see that.

In this case we have a much bigger problem. I can understand that this is a signal handler and a blocking fd may well end up hanging the process, but there does not appear to be any way to make your notification pipe work reliably. You will end up with signal events randomly not arriving.

I am, in fact, doing this because I'm having problems with tmux leaving a quickly exiting process a zombie, missing a SIGCHLD notification somehow. There did not appear to be a problem on their end, as the signal event handler should have cleaned it up. Which brought me to your signal event source where I noticed this place where a signal event could get dropped.

So what do you want to do here? Lose an event or possibly hang the process?

Reply:
Another solution would be to increment ncaught[sig] in the signal handler instead of evsig_cb, to make the byte written to the pipe just a "signal caught" notification. evsig_cb already reads all of them first and then sends signal events after. This way, if the pipe is empty, it would not block, and if it is full then EAGAIN would not matter.

Incidentally, evsig_cb sends only one event per signal, no matter how many of each were caught. Is that intentional? It would be all too easy to assume that there would be an event for each SIGCHLD.

Reply:
errno is only set to provide a descriptive message for event_warnx if write returns 0. Not setting it will result in "No error" message because write does not set errno when returning 0. EAGAIN will print "Try again" message. ENOSPC will print "No space left on device" message, which seems the most appropriate description of what happened. Other than setting this text of the logged warning, the value in errno has no further effect.

Reply:
But it is unnatural for developers who knows that most of the sockets in libevent are non blocking, I would prefer to set it to EAGAIN.

Actually the error will be printed only in case of error, and in this case errno will be set by write(2) anyway, so this errno adjustment looks superfluous.

Reply:
I'm tired of arguing about this. Commit 928016d78a5a062145b555c55269b05c0775021c removes presetting errno.

Reply:
OK, I've set errno to EAGAIN and copied your comment